physical/postgresql/postgresql.go

```diff
@@ -96,10 +159,25 @@ func NewPostgreSQLBackend(conf map[string]string, logger log.Logger) (physical.B
 		get_query:    "SELECT value FROM " + quoted_table + " WHERE path = $1 AND key = $2",
 		delete_query: "DELETE FROM " + quoted_table + " WHERE path = $1 AND key = $2",
 		list_query: "SELECT key FROM " + quoted_table + " WHERE path = $1" +
-			"UNION SELECT DISTINCT substring(substr(path, length($1)+1) from '^.*?/') FROM " +
-			quoted_table + " WHERE parent_path LIKE $1 || '%'",
+			" UNION SELECT DISTINCT substring(substr(path, length($1)+1) from '^.*?/') FROM " + quoted_table +
+			" WHERE parent_path LIKE $1 || '%'",
+		haGetLockValueQuery:
+		//only read non expired data
+		" SELECT ha_value FROM " + quoted_ha_table + " WHERE NOW() <= valid_until AND ha_key = $1 ",
+		haUpsertLockIdentityExec:
+		// $1=identity $2=ha_key $3=ha_value $4=TTL in seconds
+		//update either steal expired lock OR update expiry for lock owned by me
+		" INSERT INTO " + quoted_ha_table + " as t (ha_identity, ha_key, ha_value, valid_until) VALUES ($1, $2, $3, NOW() + $4 * INTERVAL '1 seconds'  ) " +
+			" ON CONFLICT (ha_key) DO " +
+			" UPDATE SET (ha_identity, ha_key, ha_value, valid_until) = ($1, $2, $3, NOW() + $4 * INTERVAL '1 seconds') " +
+			" WHERE (t.valid_until < NOW() AND t.ha_key = $2) OR " +
```

Comment:
Does this definitely work? (I don't know enough about Postgres isolation levels to say.)  If the lock is expired and two processes both see this condition and try to update, will one be rerun and (correctly) not update (i.e. rows affected == 0)?

Reply:
I know a lot about PG but I don't have any real experience with ON CONFLICT.  Here's my take on it nonetheless.

From the PG docs:
> ON CONFLICT DO UPDATE guarantees an atomic INSERT or UPDATE outcome; provided there is no independent error, one of those two outcomes is guaranteed, even under high concurrency. This is also known as UPSERT — "UPDATE or INSERT".

The same `ha_key` value is used in all instances ("core/lock").
 
Scenarios:
* Single process starting up sees old expired row: ON CONFLICT applies because of unique index on `ha_key`, UPDATE WHERE clause matches existing row because `valid_until < NOW()`, process grabs lock.
* Two processes starting up with empty table: one will insert, second one will not match either of the WHERE OR branches because it has a different identity and `valid_until is > NOW()`, so it will see 0 rows affected by the upsert.
* Two processes attempt to lock, expired row in table: one will be first and its WHERE clause will match due to `valid_until < NOW()`, second one will again see 0 rows affected.
* Two processes attempt to lock, unexpired row in table owned by one of them: owning process will match due to `t.ha_identity = $1 AND t.ha_key = $2` so row will be updated to have a newer `valid_until`, second one will again see 0 rows affected.

Reply:
I think we are safe. I agree with ncabatoff reasoning. Upsert is stong here since it is a single SQL command. Either you wrote it or you didnt, both cannot.  Without out it having to use two statements insert/update in would become important and require high isolation level i.e. serialized.